### ISSUE TITLE
Experiment: External Dependency Solver

### ISSUE THREAD

**Issue by lorenzleutgeb** [composer/composer#6685]
Hi!

I just ran a quick experiment and would like to get your opinion.

We're seeing quite long dependency resolution times with Composer. As dependency resolution is hard, that's not something I want to complain about. But I wanted to see how Composer's solver compares to other SAT solvers. So what I did was [a quick hack](https://github.com/composer/composer/compare/master...lorenzleutgeb:extsat) so that Composer's solver writes the formula it is solving to a file in [DIMACS format](https://logic.pdmi.ras.ru/~basolver/dimacs.html).

With this I could get a feel for the size of Composer SAT formula:

```bash
$ head -n1 composer.cnf
p cnf 11253 390122
$ ls -sh composer.cnf 
6.5M composer.cnf
```

It took Composer around 278.56 seconds to find a satisfying assignment for that formula. Now I could try and compare it against some other solver. [Plingeling](http://fmv.jku.at/lingeling/) happily solves the formula within just 0.44 seconds. That's a factor of 633.

I also tried unsatisfiable requirements, and sure enough, Plingeling reported unsatisfiability.

So it seems that there is quite some gap between the performance of Composer and Plingeling. This is what sparks my interest. Transforming a satisfying assignment back into Composer's preferred format should not be that big of a deal. What is tricky when working with an external solver is resolution of conflicts. It is not straight forward to export the state of an external solver, import it into Composer and then resolve conflicts based on that ... I'm not saying it's not possible, but I just don't know the nitty gritty details of how Composers solver does that internally.

So my proposal is as follows:
 * Only use an external solver when the size of the SAT formula (in terms of literals and clauses) exceeds some threshold. For easy formulas it's not worth the fuss.
 * Allow Composer to detect external SAT solvers (some binary from a whitelist in `$PATH`). If there is one, start it.
 * Now the external solver could either run in parallel with the Composer solver, or before the internal one. The problem is that the satisfiability of the requirements is unclear (my guess is that requirements are much more often satisfiable than unsatisfiable), so there should be some simple heuristic either running the solvers in parallel or opting for the external one first, falling back to internal in case of conflicts.
   * Assume that the requirements are satisfiable. Then it would probably make sense to just run the external solver and read the solution.
   * Assume that the requirements are not satisfiable. Then dependency resolution must be performend by the Composer solver, so starting an external one is pretty much useless.

What's your opinion on this? Note that the mechanism I sketched here can be implemented completely optional and transparent to the user! Those that know the pain of slow resolution might opt to install a SAT solver (I imagine a friendly message pointing to a tutorial in case dependency resolution takes longer than two minutes), while others might not bother and just use Composers internal solver.

**Edit:** Apart from complications with unsatisfiable requirements, I just realized that solver behaviour can be adjusted by changing [`Policy`](https://github.com/composer/composer/blob/master/src/Composer/DependencyResolver/PolicyInterface.php). As far as I see, only [`DefaultPolicy`](https://github.com/composer/composer/blob/master/doc/dev/DefaultPolicy.md) with parameters `$preferStable` and `$preferLowest` is used. In order to delegate dependency resolution to external SAT solvers, these preferences must be explicitly encoded into the SAT formula.

**Comment by naderman** [comment]
Hey, this looks really interesting. Can you share how you were measuring how long it took Composer to find the solution? Due to lazy loading even the generation of rules still results in network traffic, so depending on where or what you measured that may have severely impacted your numbers.

**Comment by lorenzleutgeb** [comment]
Please refert to [the diff I linked in the original post](https://github.com/composer/composer/compare/master...lorenzleutgeb:extsat#diff-57acde011ed9b5a18bb373af9d616bf1R229). I used what was already there :)

**Comment by naderman** [comment]
Ah indeed, not sure how I missed that, that's really just the SAT solver only then, perfect :-)

I'd also be really curious to see your input in this case but I imagine it's some real project with packagist.org dependencies? We're really missing test cases that actually take that long to solve and don't rely on network etc, so we can make them part of the test suite.

Off the top of my head I'm not sure how we could possibly encode the choices the policy makes in the SAT problem itself, do you have any idea how to accomplish this? Is there a way to somehow define a preferences for any free choices in Plingeling, or is there another suitable sat solver that has an easy way to do this?

Basically we're really not just solving the SAT question here, but also looking for a solution to the specific problem that is optimal regarding some metric of having as many packages as possible at the newest possible version.

**Comment by lorenzleutgeb** [comment]
Unfortunately I cannot share the original `composer.json` with you, [just the formulas](https://github.com/composer/composer/files/1306507/composer-cnf.zip). We use Packagist and a self-hosted repository.

Indeed, encoding these rules is not straightforward. By using policies to handle free choices you are essentially doing optimization. Plain SAT is a decision problem, not an optimization problem.

Possible approaches are:
 * Complicate your rule generation algorithm(s) so that they account for optimization, i.e. only the optimal solution is satisfying. Your policy files would have to be translated to be plugged into rule generation, not evaluation.
 * Switch to [Max-SAT](https://en.wikipedia.org/wiki/Maximum_satisfiability_problem) (or probably PMax-SAT). Hard requirements then translate to the partial SAT part that has to be satisfied and optimizations such as "have as many packages as possible at the newest possible version" would map to "soft clauses" (in PMax-SAT as many soft clauses as possible should be satisfied).
 * Switch to a different paradigm, e.g. [ASP](https://en.wikipedia.org/wiki/Answer_set_programming).

All of these options require significantly more work than I initially thought. My assumption that your `runSat` is a plain SAT solver was wrong.

Anyway, I would be happy to experiment further. In order to allow for custom policies, my idea would be to write an alternative dependency resolution "backend" based on ASP. This would work as follows:
 * Composer fetches all the data needed to build up depdency metadata, just as it is doing now. That data is then not translated into a SAT formula, but into an ASP program. Just off the top of my head this would look like:
```
% Packages 1 and 2 are available in version 1.0.0
major(1, 1). minor(1, 0). patch(1, 0).
major(2, 1). minor(2, 0). patch(2, 0).

% Major version 1 of package 1 requires major version 2 of package 3.
major(3, 1) :- major(1, 1).

% Major version 1 of package 1 conflicts with major version 1 of package 2.
:- major(2, 1), major(1, 1).
```
 * Your policies would then translate to optimization rules. They'd need to be written in ASP, but I guarantee they'll be easy to understand and very short.
 * Facts and optimization rules are concatenated and input into an ASP solver.
 * The result is mapped back into Composer's data structures.

I have no clue how to handle conflicts to be honest, but let's just ignore that for a moment. Again, I'd do this in a way that is transparent and optional.

This new dependency resolution backend would activate if the ASP solver is found on the `$PATH`, and operate transparent to the user.

**Comment by naderman** [comment]
I really have no attachment to using a SAT solver if we can get something that works faster. However I'm also not really willing to invest a huge amount of time into this experiment currently. I'd definitely be happy to help you with explaining any composer internals or discussing options if you want to give building this a try?

**Comment by lorenzleutgeb** [comment]
No need to, I think I'll manage. However, assuming that this will indeed lead to a performance gain, it would be nice get it upstream, provided that the changes will be backward compatible and transparent. :)

**Comment by naderman** [comment]
I think the main issue will be to make sure the output of the two algorithms is really identical, because otherwise you will end up with unstable results of running composer update which will lead to usability issues for developers.

**Comment by lorenzleutgeb** [comment]
True, this should be tested for.

**Comment by lorenzleutgeb** [comment]
Some notes:

 1. The [Common Upgradeability Description Format (CUDF)](http://www.mancoosi.org/cudf/) defines a format for package upgrade problems that is used to interface with some solvers. One issue with the format is that it only supports natural numbers as versions. Encoding semantic versions as `major * 10^(2*n) + minor * 10^(1*n) + patch * 10^(0*n)` would be possible for some suitable `n` but this is certainly not desired and breaks support for [pre-release versions](http://semver.org/#spec-item-9). CUDF might be extended for semantic versioning.
 2. A solver implementation for CUDF on top of the ASP system [`clingo`](https://potassco.org/clingo/) is [`aspcud`](https://potassco.org/aspcud/). It is nicely explained in [*aspcud: A Linux Package Configuration Tool Based on Answer Set Programming*](https://arxiv.org/abs/1109.0113).
 3. The [Advanced Package Tool (APT)](https://wiki.debian.org/Apt) project has defined an [External Dependency Solver Protocol (EDSP)](https://github.com/Debian/apt/blob/master/doc/external-dependency-solver-protocol.txt) which specifies interaction between a package manager (in this case APT) and an external solver. Composer might implement a similar protocol.

@zacchiro I would appreciate if you could comment on possible extensions of CUDF for [semantic versioning](http://semver.org/), i.e. lifting versions from integers to tuples with a peculiar interpretation. Do you know of any work that has considered this? What is your opinion on the matter? Do you think that this would make sense? Thank you.

**Comment by zacchiro** [comment]
Just to recap for those who might not be aware of it, the expectation when using a CUDF solver is that upgrade scenarios will be preprocessed totally sorting all versions and mapping them to integers; and doing the same gymnastic back when interpreting the result given by the solver.

I've just checked semver again (I hadn't in a while) and it seems to support now pre-release version components, which possibly wasn't the case when CUDF was designed. That's nice!

There is no profound reason why CUDF couldn't use semver as version identifiers instead of integers, but IME that will not solve the *general* need of having to do preprocessing, because "real" version numbers out there contains all sort of horrible things:

1. pre-release version identifiers alpha/beta/etc (which are indeed supported nicely by semver now)
2. epochs, to start new "generations" of version numbers: 2:1.2.3 > 1:3.4.5 > 4.5.6 (often used when upstreams change versioning schemes, e.g., adding/removing dates from their version numbers)
3. Debian-style "~" which looks like a general form of (1), which I'm not sure can be fully expressed in it

So while I'm not opposed to consider "upgrading" the CUDF spec to support semver natively, I'm confident there will still be package/component ecosystems where you'll have to do pre/postprocessing on versions anyhow. But it will indeed benefit ecosystems that already enforce semver, and it won't be a regression as if you're mapping to integers anyhow, they will still be valid semver versions.

And of course existing CUDF solvers will need to be ported to this potential change in the spec before being able to be used natively with semver versions.

**Comment by lorenzleutgeb** [comment]
OK, thank you! I missed that preprocessing is expected here.

**Comment by zacchiro** [comment]
To be clear: preprocessing in the pipeline is not a goal per se (of course :)). But it's needed to address all the craziness in version numbering that exists out there. If you're in a context in which you can *guarantee* that all version numbers adhere to semver versioning in the strict sense of their lexical conventions, and if we semver support to the CUDF spec, than preprocessing won't be needed in your case.

With my CUDF spec editor hat on, I'd be open to consider adding semver support if various contexts in which semver is strictly enforced exist and can benefit from this change.

**Comment by naderman** [comment]
Not sure exactly where this is going, but this will have to work with exactly the versions composer accepts, and the semver library included in it already takes care of sanitization. Each package object has a normalized version number at the time the solver needs to deal with it. Further the current solver solution simply has a first step that deals with turning each individual version into a literal and creates the necessary rules (conjunctions) depending on version numbers so the solver itself just gets a bunch of conjunctions.

**Comment by naderman** [comment]
Sadly I think resources aren't going to permit to finish this for 2.0, so moving it off the core milestone.

**Comment by pinkeen** [comment]
I made a PoC that integrates, slightly modified MiniSat via FFI. It does not take policies into account though.

I've made it almost work correctly for the default case where you want the latest versions. I've noticed that the lower the variable number the higher the version and modified it to choose the lowest one when picking the branching literal. It works for root packages but doesn't always choose the latest dep versions. I wonder if this approach makes sense...

The perf difference is similar to what OP reported. I am even more interested in the memory usage decrease because we have projects which regularly require 1-3G for generating a lock file.

**Comment by pinkeen** [comment]
BTW I think FFI is the way to go here. If php < 7.4 or no bundled binary is supported then it can always fall back to internal solver.

The stripped lib is ~60kB and I can cover 99% of OSX, Linux, windows systems with 4-5 binaries.

**Comment by pinkeen** [comment]
I've looked at MiniSat deeper and it seems I've approached it the wrong way. It's meant to be customized and provides mechanisms to do that (as stated in the [paper](http://minisat.se/downloads/MiniSat.pdf)). 

It should be relatively straightforward to extend it with custom policy operating on extra data (e.g. package stability). Effectively making it almost a native reimplementation of the internal solver. You could not plug other solvers this way, but  I don't see a reason to with the perf gains provided by this impl.

**Comment by lorenzleutgeb** [comment]
Sweet!

Linking against one solver definitely is an option.
However, IMO the more portable way still is to rely on standard formats (like DIMACS, as mentioned, or [SMT-LIB](http://smtlib.cs.uiowa.edu/)) as intermediate step.

In particular, I think that the richer SMT-LIB language would make it easier to express policies. The input for the solver could look much more human readable, with dedicated predicates that express e.g. SemVer version precedence.

I experimented a bit with SMT and SemVer (in generel, not directly related to Composer), but didn't push to GitHub yet.

**Comment by pinkeen** [comment]
Thanks for the reply!

> However, IMO the more portable way still is to rely on standard formats (like DIMACS, as mentioned, or [SMT-LIB](http://smtlib.cs.uiowa.edu/)) as intermediate step.

According to my understanding DIMACS and formats which encode only the boolean clauses are not option. As mentioned previously in this conversation finding a solution is not enough. We need to find **the optimal** solution given our problem domain, which would be the latest/oldest (stable) packages (satisfy the DefaultPolicy). I don't see a way to express the policy as boolean clauses, as it's not a constraint but rather an optimization.

Having a universal interface with pluggable implementations was my first approach, however, I am not sure how feasible (in terms of initial and maintenance work needed) this is, especially given that we need ensure result stability.

The CUDF and SMT-LIB look interesting. Personally I would lean towards something that is designed to resolve package deps. I would also take a look at integrating libzypp or libsolv directly.

**Comment by Seldaek** [comment]
> The perf difference is similar to what OP reported. I am even more interested in the memory usage decrease because we have projects which regularly require 1-3G for generating a lock file.

Is that using composer v1 or v2? Because at this point (and with further improvements coming in 2.1) IMO the situation got so much better that it's less compelling to work on a C/FFI implementation. 

That said if someone comes up with one that is compatible with DefaultPolicy etc and transparently boosts performance that'd for sure sound interesting.

**Comment by pinkeen** [comment]
> > The perf difference is similar to what OP reported. I am even more interested in the memory usage decrease because we have projects which regularly require 1-3G for generating a lock file.
> 
> Is that using composer v1 or v2? Because at this point (and with further improvements coming in 2.1) IMO the situation got so much better that it's less compelling to work on a C/FFI implementation.

Do you mean memory complexity and the improvements coming for MultiConflictRule?

I have to retest, because we still use 1.x internally (as 2.x is not supported yet by some plugins we require) and I've used it for the first tests. I've based my PoC on current master though so I will check this shortly.

**Comment by pinkeen** [comment]
@Seldaek I've ran some comparison tests between v2 and my current PoC on one of the prolematic projects.

Composer v2 wins hands-down! 

I made an error - my initial numbers were incorrect because I did realize I need to transform the MultiConflict rules into combinations - that's where I got down to fraction of a second vs ~140s for composer v1) 

Now that minisat needs to crunch all those split conflict rules while composer doesn't need to the tables have turned - it's around 40s for composer v2 vs 140s for minisat. Also composer's memory footprint has decreased to 300-400M which is entirely acceptable. Good job!

I've just completely lost interest in this. At least I've learned a thing or two about SAT solvers and composer's internals so maybe all this time won't go to waste ;)

**Comment by Seldaek** [comment]
Hehe ok yeah that is what I suspected. We might gain some extra memory by moving all the rules and the solver to C, but that's a huge amount of work IMO for possibly little gain. Especially if all the improvements lined up for 2.1 pan out, then memory usage will be reduced quite a bit again and make this even less compelling.

**Comment by Seldaek** [comment]
Also.. you mentioned you are still stuck on Composer 1 due to plugins, which are these? Anything we can do to help nudge things forward?

**Comment by lorenzleutgeb** [comment]
FWIW, I pushed my scratches about using SMT: https://github.com/lorenzleutgeb/composer-smt/blob/main/semver.smt

The main problem obstacle being that I don't know enough about Composer.

**Comment by pinkeen** [comment]
> Also.. you mentioned you are still stuck on Composer 1 due to plugins, which are these? Anything we can do to help nudge things forward?

This should be a straightforward migration, but thanks!

**Comment by stof** [comment]
@Seldaek @naderman with the PoolOptimizer we now have, I think this should be closed.

**Comment by naderman** [comment]
I suppose so. Still be an interesting project to play around with some external solvers and also see if optimizations can be improved in others, but considering how little a problem this still is in practice now, I don't think we'll put any resources into this.
